### ISSUE TITLE
[0.9.3-rc1] Out of memory and influxd was killed by OS

### ISSUE THREAD

**Issue by maxone** [influxdata/influxdb#3728]
* OS
  CentOS release 6.6 Linux version 2.6.32-504.16.2.el6.x86_64

* Database
```
create database hawk
create retention policy "10y" on hawk duration 36000d replication 1 default
show retention policies on hawk
```

* Data point 
```
2007-01-01 12:00:00 field1=100,field2=200, application=account0,host=host0
```
* InfluxDB 0.9.2.1 Single-node
 1.  30 million data was successfully batch wrote to 10 tables with 2 tags but memory was highly occupied

 2.  OOM occurred and "influxDB process" was killed by OS when another 30 million data was batch wrote to 10 new tables with 2 different  tags after step 1

<img width="870" alt="screen shot 2015-08-19 at 8 09 06 pm" src="https://cloud.githubusercontent.com/assets/2100584/9355677/ce633d42-46ad-11e5-86bd-612f3384f3ae.png">

<img width="970" alt="screen shot 2015-08-19 at 7 30 06 pm" src="https://cloud.githubusercontent.com/assets/2100584/9355473/3aae8724-46ac-11e5-8439-c74f913f43b2.png">

* InfluxDB 0.9.3rc Single-node
  1. with same steps, I got exception in log file and the influxd crashed
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x8 pc=0x44219f]

goroutine 9938 [running]:
github.com/influxdb/influxdb/tsdb.(*Shard).validateSeriesAndFields(0xc3b61f3600, 0xc26a37e000, 0xb56, 0xe00, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /opt/cockroach/data/src/github.com/influxdb/influxdb/tsdb/shard.go:336 +0x11b7
github.com/influxdb/influxdb/tsdb.(*Shard).WritePoints(0xc3b61f3600, 0xc26a37e000, 0xb56, 0xe00, 0x0, 0x0)
        /opt/cockroach/data/src/github.com/influxdb/influxdb/tsdb/shard.go:154 +0x75
github.com/influxdb/influxdb/tsdb.(*Store).WriteToShard(0xc2080341a0, 0x2e, 0xc26a37e000, 0xb56, 0xe00, 0x0, 0x0)
        /opt/cockroach/data/src/github.com/influxdb/influxdb/tsdb/store.go:289 +0xf7
github.com/influxdb/influxdb/cluster.func·002(0x2e, 0x1, 0xc26a37e000, 0xb56, 0xe00)
        /opt/cockroach/data/src/github.com/influxdb/influxdb/cluster/points_writer.go:241 +0xfc
created by github.com/influxdb/influxdb/cluster.(*PointsWriter).writeToShard
        /opt/cockroach/data/src/github.com/influxdb/influxdb/cluster/points_writer.go:271 +0x30b

goroutine 1 [chan receive, 1 minutes]:
main.(*Main).Run(0xc20806bb00, 0xc20800a010, 0x4, 0x4, 0x0, 0x0)
        /opt/cockroach/data/src/github.com/influxdb/influxdb/cmd/influxd/main.go:91 +0x777
main.main()
        /opt/cockroach/data/src/github.com/influxdb/influxdb/cmd/influxd/main.go:42 +0xdc

goroutine 6 [syscall, 1 minutes]:
os/signal.loop()
        /usr/lib/golang/src/os/signal/signal_unix.go:21 +0x1f
created by os/signal.init·1
        /usr/lib/golang/src/os/signal/signal_unix.go:27 +0x35

goroutine 17 [syscall, 1 minutes, locked to thread]:
runtime.goexit()
        /usr/lib/golang/src/runtime/asm_amd64.s:2232 +0x1

```

<img width="879" alt="screen shot 2015-08-19 at 8 06 02 pm" src="https://cloud.githubusercontent.com/assets/2100584/9355631/622c70f8-46ad-11e5-8dfe-a93014506b1d.png">

**Comment by mjdesa** [comment]
@maxone can you send me an example data point in the line protocol format.

**Comment by maxone** [comment]
@mjdesa 
```
jvm1,application=demo1,host=bc-vm1 fgc=6,fgct=8,mm=3555,om=1277,thread_count=8,ygc=7,ygct=8,ym=7209 1440033471224
jvm1,application=demo1,host=bc-vm1 fgc=4,fgct=7,mm=1580,om=3542,thread_count=5,ygc=8,ygct=2,ym=5796 1440033531224
jvm1,application=demo1,host=bc-vm1 fgc=9,fgct=2,mm=9032,om=8589,thread_count=11,ygc=8,ygct=3,ym=2174 1440033591224
jvm1,application=demo1,host=bc-vm1 fgc=6,fgct=3,mm=8255,om=6777,thread_count=94,ygc=9,ygct=2,ym=9363 1440033651224
jvm1,application=demo1,host=bc-vm1 fgc=4,fgct=1,mm=7139,om=1296,thread_count=88,ygc=3,ygct=0,ym=8831 1440033711224
jvm1,application=demo1,host=bc-vm1 fgc=8,fgct=4,mm=6784,om=6427,thread_count=63,ygc=8,ygct=2,ym=8452 1440033771224
jvm1,application=demo1,host=bc-vm1 fgc=8,fgct=0,mm=9435,om=5189,thread_count=37,ygc=4,ygct=1,ym=1872 1440033831224
jvm1,application=demo1,host=bc-vm1 fgc=5,fgct=7,mm=837,om=108,thread_count=34,ygc=3,ygct=2,ym=6874 1440033891224
jvm1,application=demo1,host=bc-vm1 fgc=6,fgct=7,mm=5479,om=5790,thread_count=20,ygc=4,ygct=1,ym=5988 1440033951224
jvm1,application=demo1,host=bc-vm1 fgc=6,fgct=3,mm=8629,om=1357,thread_count=70,ygc=7,ygct=7,ym=9671 1440034011224
jvm1,application=demo1,host=bc-vm1 fgc=4,fgct=4,mm=7357,om=4667,thread_count=38,ygc=1,ygct=4,ym=4651 1440034071224
jvm1,application=demo1,host=bc-vm1 fgc=8,fgct=2,mm=4039,om=7918,thread_count=6,ygc=7,ygct=0,ym=3838 1440034131224
```

**Comment by mjdesa** [comment]
@maxone how much RAM does the machine you're running this on have?

**Comment by maxone** [comment]
@mjdesa 60GB

**Comment by ccollicutt** [comment]
This happens with me too. Ubuntu 14.04, in an lxc container (though only one on the physical host), 32gb of memory. influxdb eats up all memory and swap and is killed.

**Comment by pauldix** [comment]
Are you issuing queries at the same time? Is this only for writes? What's the cardinality of the number of series you have? (that's measurement + unique tagset). How are you writing the data in?

I can write 30m data points across 100k series in seconds and fill up less than a few GB of memory.

**Comment by beckettsean** [comment]
@maxone @ccollicutt how are you writing points to the database? Are you batching? Using HTTP? UDP?

**Comment by ccollicutt** [comment]
For me, http via telegraf for metrics.

**Comment by beckettsean** [comment]
Thanks for the quick answer, @ccollicutt. Are you also running an 0.9.3 variant?

**Comment by maxone** [comment]
Http batching.

**Comment by ccollicutt** [comment]
@beckettsean am on 0.9.3 now, seems better, but I had to remove a lot of data for other reasons. But so far seems better.

**Comment by KiNgMaR** [comment]
I'm seeing the same or a similar issue on CentOS 6.6 (16 GB RAM, 8 CPU cores). I tested with 0.9.3 and 0.9.4-nightly. After 1-2 hours InfluxDB memory usage hits 16 GB.

I think it may be related to continuous queries or at least accelerated by them. My write load is as follows:

    [wal] 2015/08/28 14:01:06 Flushing 0 measurements and 211 series to index
    [wal] 2015/08/28 14:01:10 Flushing 0 measurements and 320984 series to index
    [wal] 2015/08/28 14:11:06 Flushing 0 measurements and 201 series to index
    [wal] 2015/08/28 14:11:10 Flushing 0 measurements and 240350 series to index
    [wal] 2015/08/28 14:21:06 Flushing 0 measurements and 27 series to index
    [wal] 2015/08/28 14:21:09 Flushing 0 measurements and 185272 series to index
    [wal] 2015/08/28 14:31:06 Flushing 0 measurements and 162 series to index
    [wal] 2015/08/28 14:31:09 Flushing 0 measurements and 166390 series to index
    [wal] 2015/08/28 14:41:06 Flushing 0 measurements and 129 series to index
    [wal] 2015/08/28 14:41:08 Flushing 0 measurements and 173562 series to index
    [wal] 2015/08/28 14:51:06 Flushing 0 measurements and 532 series to index
    [wal] 2015/08/28 14:51:10 Flushing 0 measurements and 148094 series to index
    [wal] 2015/08/28 15:01:06 Flushing 0 measurements and 112 series to index
    [wal] 2015/08/28 15:01:06 Flushing 0 measurements and 46912 series to index
    [wal] 2015/08/28 15:11:06 Flushing 0 measurements and 115 series to index

All measurements have a single column. Timestamps are never sent by clients but always added by the InfluxDB server. Data is being sent using UDP and the following settings:

    batch-size = 10000
    batch-timeout = "5s"

The WAL is configured like this:

    wal-dir = "/var/opt/influxdb/wal"
    wal-enable-logging = true
    wal-ready-series-size = 12800
    # wal-compaction-threshold = 0.6
    wal-max-series-size = 1048576
    wal-flush-cold-interval = "15m"
    wal-partition-size-threshold = 20971520

I can reproduce this with a fresh database (z1 as I presume). I have three very simple CQs set up, all with 1 tag and group by time (1m).

After stopping all writes, memory usage is coming down, but slowly. After 20 minutes it went from 17 GB RSS to 10 GB.

**Comment by DanielMorsing** [comment]
One of the things that's useful for hunting down these issues is a memory profile to see where the memory is going. Could you try running `go tool pprof -proto -output /tmp/memprofile http://serverhostname:8086/debug/pprof/heap` and sending the `/tmp/memprofile` file?

**Comment by svenmueller** [comment]
we are also running into this error (ubuntu 14.04, 4GB node)

**Comment by ankanm** [comment]
We're seeing something similar on ubuntu (14.04, 128gb).
The following repeats for quite a while
```
[monitor] 2016/01/16 19:43:42 failed to store statistics: timeout
[collectd] 2016/01/16 19:43:43 failed to write point batch to database "metrics_db": timeout
[collectd] 2016/01/16 19:43:48 failed to write point batch to database "metrics_db": timeout
```
before finally -
```
runtime: out of memory: cannot allocate 1048576-byte block (137438035968 in use)
fatal error: out of memory

runtime stack:
runtime.throw(0xfe5baa)
        /root/.gvm/gos/go1.4.2/src/runtime/panic.go:491 +0xad
runtime.largeAlloc_m()
        /root/.gvm/gos/go1.4.2/src/runtime/malloc.c:390 +0xee
runtime.onM(0x1005d70)
        /root/.gvm/gos/go1.4.2/src/runtime/asm_amd64.s:257 +0x68
runtime.mstart()
        /root/.gvm/gos/go1.4.2/src/runtime/proc.c:818
...
```

**Comment by toddboom** [comment]
@ankanm what version of InfluxDB were you running?

**Comment by ankanm** [comment]
@toddboom we're on influxdb 0.9.5.1.

**Comment by toddboom** [comment]
These issues should be resolved on the most recent 0.10.x releases.
